engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.swift

```diff
@@ -0,0 +1,159 @@
+      retain: nil,
+      release: nil,
+      copyDescription: nil
+    )
+    guard
+      let createdTimer = CFRunLoopTimerCreate(
+        kCFAllocatorDefault,
+        CFAbsoluteTime.greatestFiniteMagnitude,
+        CFAbsoluteTime.greatestFiniteMagnitude,
+        0,
+        0,
+        { timer, info in
+          let runner = Unmanaged<FlutterRunLoop>.fromOpaque(info!).takeUnretainedValue()
+          runner.performExpiredTasks()
+        },
+        &timerContext
+      )
+    else {
+      fatalError("Failed to create CFRunLoopTimer")
+    }
+    timer = createdTimer
+    CFRunLoopAddTimer(runLoop, timer, .commonModes)
+    CFRunLoopAddTimer(runLoop, timer, Self.flutterRunLoopMode)
+  }
+
+  deinit {
+    CFRunLoopTimerInvalidate(timer)
+    CFRunLoopRemoveTimer(runLoop, timer, .commonModes)
+    CFRunLoopRemoveTimer(runLoop, timer, Self.flutterRunLoopMode)
+    CFRunLoopSourceInvalidate(source)
+    CFRunLoopRemoveSource(runLoop, source, .commonModes)
+    CFRunLoopRemoveSource(runLoop, source, Self.flutterRunLoopMode)
+  }
+
+  // Ensures that the `FlutterRunLoop` for main thread is initialized. Only
+  // needs to be called once and must be called on the main thread.
+  @objc public static func ensureMainLoopInitialized() {
+    assert(Thread.isMainThread, "Must be called on the main thread.")
+    if _mainRunLoop == nil {
+      _mainRunLoop = FlutterRunLoop()
+    }
+  }
+
+  // The `FlutterRunLoop` for the main thread.
+  @objc public static var mainRunLoop: FlutterRunLoop {
+    assert(
+      _mainRunLoop != nil,
+      "Main run loop has not been initialized. Call ensureMainLoopInitialized() first."
+    )
+    return _mainRunLoop!
+  }
+
+  // Schedules a block to be executed on the main thread.
+  @objc(performBlock:afterDelay:)
+  public func perform(block: @escaping () -> Void, afterDelay delay: TimeInterval) {
+    tasksLock.lock()
+    defer { tasksLock.unlock() }
+
+    tasks.append(Task(block: block, targetTime: CFAbsoluteTimeGetCurrent() + delay))
+    if delay > 0 {
+      rearmTimer()
+    } else {
+      CFRunLoopSourceSignal(source)
+      CFRunLoopWakeUp(runLoop)
+    }
+  }
+
+  // Schedules a block to be executed on the main thread after a delay.
+  @objc(performBlock:)
+  public func perform(block: @escaping () -> Void) {
+    perform(block: block, afterDelay: 0)
+  }
+
+  private func performExpiredTasks() {
+    var expiredTasks: [Task] = []
+
+    tasksLock.lock()
+    let now = CFAbsoluteTimeGetCurrent()
+    tasks.removeAll { task -> Bool in
+      if task.targetTime <= now {
+        expiredTasks.append(task)
+        return true
+      }
+      return false
+    }
+    rearmTimer()
+    tasksLock.unlock()
+
+    for task in expiredTasks {
+      task.block()
+    }
+  }
+
+  private func rearmTimer() {
+    var nextFireTime: CFAbsoluteTime = .greatestFiniteMagnitude
+    for task in tasks {
+      nextFireTime = min(nextFireTime, task.targetTime)
+    }
```

Comment:
nit: can do `min` to avoid `var`
```
let nextTask = tasks.min { $0.targetTime < $1.targetTime }
let nextFireTime = nextTask.targetTime ?? .greatestFiniteMagnitude 
```

Reply:
Ah good point. Done using reduce instead.

Reply:
nice! `min` is basically a `reduce` (with initial value of `nil`), but swift implemented it for you.